### PR TITLE
Fix MRI Ruby vs. JRuby XML child namespace output differences

### PR DESCRIPTION
**Pull request by johnnyshields**
Fixes https://github.com/sparklemotion/nokogiri/issues/3455

### CODE REVIEW

**Comment by johnnyshields** [comment]
@flavorjones please approve CI, thanks!

**Comment by flavorjones** [comment]
I'm not opposed to this approach. I'm wondering if the namespace stack should be managed in the `getAttrsAndNamespaces()` function, though, since that's already managing one namespace stack (for canonicalized output). WDYT?

**Comment by johnnyshields** [comment]
@flavorjones I think you're right, I'll look more into getAttrsAndNamespaces tomorrow

**Comment by flavorjones** [comment]
Thank you so much for diving in and drafting this pull request! I sincerely wish there were more folks willing to contribute to the JRuby implementation. ❤️

**Comment by johnnyshields** [comment]
@flavorjones code is ready for a review. What is the best way to test this? Are there instructions on how to build JRuby locally?

**Comment by flavorjones** [comment]
@johnnyshields Existing "how to contribute" docs are at https://github.com/sparklemotion/nokogiri/blob/main/CONTRIBUTING.md

but the TLDR is `bundle install ; bundle exec rake compile test`

**Comment by johnnyshields** [comment]
@flavorjones This is ready for final review. While testing I found a separate issue unrelated to this PR: https://github.com/sparklemotion/nokogiri/issues/3457

File: test/namespaces/test_namespaces_in_child_elements.rb [hunk 1]
```diff
@@ -0,0 +1,257 @@
+# frozen_string_literal: true
+
+require 'helper'
+
+describe 'namespaces in child elements' do
+  it 'does not repeat xmlns definitions in child elements' do
+    # https://github.com/sparklemotion/nokogiri/issues/3455
+    doc = Nokogiri::XML::Builder.new do |xml|
```
Comment (flavorjones) [inline test/namespaces/test_namespaces_in_child_elements.rb:8]: I'm curious if there's a specific reason you have in mind for using Builder to construct the test subjects in this file? It complicates the tests a tiny bit, and as a reader I'm not sure whether we're testing Builder behavior or serialization behavior.

I'd prefer constructing the document by parsing a string, but I don't feel strongly about it. If you want to keep it as-is using Builder, then please just add a brief comment at the top explaining that we're testing serialization and not tree construction.

File: test/namespaces/test_namespaces_in_child_elements.rb [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+      end.doc.to_xml(save_with: Nokogiri::XML::Node::SaveOptions::AS_XML)
+
+      assert_includes(doc, '<root xmlns="http://default.org/" xmlns:ns="http://prefixed.org/">')
+      assert_includes(doc, '<default_element>in default namespace</default_element>')
+      assert_includes(doc, '<ns:prefixed_element>in prefixed namespace</ns:prefixed_element>')
+      assert_includes(doc, '<mixed xmlns="http://new-default.org/">')
+      # TODO: The JRuby behavior here is incorrect.
+      # See: https://github.com/sparklemotion/nokogiri/issues/3457
+      if Nokogiri.jruby?
+        assert_includes(doc, '<new_default xmlns="http://default.org/">in new default namespace</new_default>')
```
Comment (flavorjones) [inline test/namespaces/test_namespaces_in_child_elements.rb:194]: See earlier `pending` comment

**Review by flavorjones — CHANGES_REQUESTED** [review]
Thank you for the thorough test coverage! ❤️

File: test/namespaces/test_namespaces_in_child_elements.rb [hunk 3]
```diff
@@ -0,0 +1,257 @@
+# frozen_string_literal: true
```
[the remark this answers is not in the document]
Comment (johnnyshields) [reply test/namespaces/test_namespaces_in_child_elements.rb:1]: Test files are not present in the gemspec, at least not from what I can see. Please double-check.

File: test/namespaces/test_namespaces_in_child_elements.rb (hunk 1 above)
Comment (johnnyshields) [reply test/namespaces/test_namespaces_in_child_elements.rb:8]: It's how I am using Nokogiri for RubySaml, and I think it's also probably the most compact syntax. Added comment:

```
# This test uses Nokogiri::XML::Builder for code terseness.
# The test is primarily intended to test serialization behavior,
# not tree construction.
```

File: test/namespaces/test_namespaces_in_child_elements.rb (hunk 2 above)
Comment (johnnyshields) [reply test/namespaces/test_namespaces_in_child_elements.rb]: Done.

File: test/namespaces/test_namespaces_in_child_elements.rb [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+        end
+      end.doc.to_xml(save_with: Nokogiri::XML::Node::SaveOptions::AS_XML)
+
+      assert_includes(doc, '<root xmlns="http://outer-namespace.org/">')
+      assert_includes(doc, '<outer>in outer namespace</outer>')
+      assert_includes(doc, '<inner xmlns="http://inner-namespace.org/">')
+      # TODO: The JRuby behavior here is incorrect.
+      # See: https://github.com/sparklemotion/nokogiri/issues/3457
+      if Nokogiri.jruby?
+        assert_includes(doc, '<element xmlns="http://outer-namespace.org/">in inner namespace</element>')
```
[the remark this answers is not in the document]
Comment (johnnyshields) [reply test/namespaces/test_namespaces_in_child_elements.rb]: Done.

File: test/namespaces/test_namespaces_in_child_elements.rb [hunk 5]
```diff
@@ -0,0 +1,257 @@
+# frozen_string_literal: true
+
+require 'helper'
```
[the remark this answers is not in the document]
Comment (johnnyshields) [reply test/namespaces/test_namespaces_in_child_elements.rb]: Done.

File: test/namespaces/test_namespaces_in_child_elements.rb [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+
+describe 'namespaces in child elements' do
```
[the remark this answers is not in the document]
Comment (johnnyshields) [reply test/namespaces/test_namespaces_in_child_elements.rb]: Done.

File: test/namespaces/test_namespaces_in_child_elements.rb [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+      end
+    end.doc.to_xml(save_with: Nokogiri::XML::Node::SaveOptions::AS_XML)
+
+    assert_includes(doc, '<dnd:adventure xmlns:dnd="http://www.w3.org/dungeons#">')
+    assert_includes(doc, '<dnd:party xmlns:dnd="http://www.w3.org/dragons#">')
+    assert_includes(doc, '<dnd:members>')
+    # TODO: The MRI (non-JRuby) behavior here is incorrect.
+    # See: https://github.com/sparklemotion/nokogiri/issues/3458
+    if Nokogiri.jruby?
+      assert_includes(doc, '<dnd:character xmlns:dnd="http://www.w3.org/dungeons#">')
```
[the remark this answers is not in the document]
Comment (johnnyshields) [reply test/namespaces/test_namespaces_in_child_elements.rb]: Done.

**Comment by johnnyshields** [comment]
@flavorjones all comments are addressed, please check.

File: test/namespaces/test_namespaces_in_child_elements.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (flavorjones) [reply test/namespaces/test_namespaces_in_child_elements.rb]: Oh damn, you're right, I forgot I removed them. 🤦

**Review by flavorjones — APPROVED** [review]
Looks good, will merge if the test suite passes.

**Comment by flavorjones** [comment]
Will get this into a bugfix release soon. But first I'd like to see if I can easily fix the other bugs we're finding and get them in, too.

**Comment by johnnyshields** [comment]
@flavorjones awesome, I really appreciate it!

**Comment by flavorjones** [comment]
@johnnyshields I really appreciate you digging in and fixing this!

**Comment by flavorjones** [comment]
Side note: I'm integration testing Nokogiri `main` against ruby-saml `main` in the "downstream" workflow (e.g. https://github.com/sparklemotion/nokogiri/actions/runs/13752072852/job/38454295028).

Should I be testing ruby-saml with JRuby as well, do you think? You're obviously finding edge cases that nobody else has in 15+ years so it might be worthwhile ...

**Comment by johnnyshields** [comment]
@flavorjones yes, please test JRuby and TruffleRuby for RubySaml if possible 🙇‍♂️

**Comment by flavorjones** [comment]
See #3465 for CI coverage

**Comment by pitbulk** [comment]
Great work here!
